tests/qos/test_buffer.py

```diff
@@ -0,0 +1,848 @@
+        pytest_assert('Profile {} has been removed from ASIC_DB'.format(sai_oid))
+
+
+def check_dynamic_th_in_appldb(duthost, profile, expected_dynamic_th, must_exist = False):
+    time.sleep(10)
+    dynamic_th = duthost.shell('redis-cli hget "BUFFER_PROFILE_TABLE:{}" dynamic_th'.format(profile))['stdout']
+    pytest_assert(not must_exist and not dynamic_th or dynamic_th == expected_dynamic_th, "dynamic_th of profile {} in APPL_DB isn't correct".format(profile))
+
+
+def fetch_initial_asic_db(duthost):
+    profiles_in_asicdb = duthost.shell('redis-cli -n 1 keys "ASIC_STATE:SAI_OBJECT_TYPE_BUFFER_PROFILE*"')['stdout']
+    return set(profiles_in_asicdb.split('\n'))
+
+
+def _compose_dict_from_cli(fields_list):
+    return dict(zip(fields_list[0::2], fields_list[1::2]))
+
+
+def check_buffer_profile_details(duthost, initial_profiles, profile_name, profile_oid, pool_oid):
+    profile_appldb = _compose_dict_from_cli(duthost.shell('redis-cli hgetall BUFFER_PROFILE_TABLE:{}'.format(profile_name))['stdout'].split('\n'))
+    logging.debug("appl db profile {}".format(profile_appldb))
+
+    # check the profile against the standard value
+    m = re.search(lossless_profile_pattern, profile_name)
+    if m:
+        # this means it's a dynamic profile
+        speed = m.group(1)
+        cable_length = m.group(2)
+        std_profiles_for_speed = default_lossless_headroom_data.get(speed)
+        std_profile = std_profiles_for_speed.get(cable_length)
+        if std_profile:
+            # this means it's a profile with std speed and cable length. we can check whether the headroom data is correct
+            pytest_assert(profile_appldb['xon'] == std_profile['xon'] and profile_appldb['xoff'] == std_profile['xoff'] and profile_appldb['size'] == std_profile['size'],
+                          "Generated profile {} doesn't match the std profile {}".format(profile_appldb, std_profile))
+        else:
+            for std_cable_len, std_profile in std_profiles_for_speed.items():
+                if int(std_cable_len[:-1]) > int(cable_length[:-1]):
+                    pytest_assert(int(std_profile['xoff']) >= int(profile_appldb['xoff']),
+                                  "XOFF of generated profile {} is greater than standard profile {} while its cable length is less".format(profile_appldb, std_profile))
+                else:
+                    pytest_assert(int(std_profile['xoff']) <= int(profile_appldb['xoff']),
+                                  "XOFF of generated profile {} is less than standard profile {} while its cable length is greater".format(profile_appldb, std_profile))
+
+    profiles_in_asicdb = set(duthost.shell('redis-cli -n 1 keys "ASIC_STATE:SAI_OBJECT_TYPE_BUFFER_PROFILE*"')['stdout'].split('\n'))
+    diff = profiles_in_asicdb - initial_profiles
+    if len(diff):
+        profile_oid = diff.pop()
+
+    logging.debug("initial profiles {} current profiles {} difference {}".format(initial_profiles, profiles_in_asicdb, diff))
```

Comment:
please consider changing
logging.debug("initial profiles {} current profiles {} difference {}".format(initial_profiles, profiles_in_asicdb, diff))
to
logging.debug("initial profiles {} and current profiles {} has the following difference(s) {}".format(initial_profiles, profiles_in_asicdb, diff))